### PR TITLE
planner: optimize NewPlanCacheKey

### PR DESCRIPTION
**Pull request by zyguan**
### What problem does this PR solve?

Issue Number: ref #54047

Problem Summary: `NewPlanCacheKey` introduce many memory allocations.

### What changed and how does it work?

Reduce the mem alloc of `NewPlanCacheKey` (as well as `key.Hash`). See the following benchmark results.

**micro benchmark**

https://clinic.pingcap.com.cn/portal/#/orgs/541/clusters/7381420967173394916?from=1718622510&to=1718623110

baseline

this PR

**sysbench: oltp point select**
  | avg lat | p95 lat | throughput | througput diff
-- | -- | -- | -- | --
baseline | 1.51 | 2.97 | 66031.42 | -
this PR | 1.48 | 2.91 | 67443.97 | 2.14%

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/plan_cache_test.go [hunk 1]
```diff
@@ -1825,3 +1826,15 @@ func TestIndexRange(t *testing.T) {
 	tk.MustQuery(`SELECT t0.* FROM t0 WHERE (id = 1 or id = 9223372036854775808);`).Check(testkit.Rows("1"))
 	tk.MustQuery("SELECT t1.c0 FROM t1 WHERE t1.c0!=BIN(-1);").Check(testkit.Rows("1"))
 }
+
+func BenchmarkPlanCacheKey(b *testing.B) {
```
Comment (winoros) [inline pkg/planner/core/plan_cache_test.go:1830]: Can we add another benchmark for the large key cases?

File: pkg/planner/core/plan_cache_utils.go [hunk 2]
```diff
@@ -229,6 +229,12 @@ func GeneratePlanCacheStmtWithAST(ctx context.Context, sctx sessionctx.Context,
 	return preparedObj, p, paramCount, nil
 }
 
+// relatedTblInfo is used to hold related table for building plan cache key.
+type relatedTblInfo struct {
+	id  int64
+	rev uint64
```
Comment (crazycs520) [inline pkg/planner/core/plan_cache_utils.go:235]:
```suggestion
	ver uint64
```

the name should be `version`?

Comment (zyguan) [reply pkg/planner/core/plan_cache_utils.go:235]: But the value comes from `TableInfo.Revision`.

**Comment by cfzjywxk** [comment]
@qw4990 @winoros 
PTAL again for the optimization logic for single table, thx

**Comment by zyguan** [comment]
> @qw4990 @winoros PTAL again for the optimization logic for single table, thx

@qw4990 have done some refactor works, this PR might need to be reconsidered.

**Pull request closed without merging**